Add GeoSpark Python to PyPi.

Imbruced opened the pull request:
## Is this PR related to a proposed Issue?
No
## What changes were proposed in this PR?
Update GeoSpark Python to 1.3.0, PyPi support 
## How was this patch tested?
None
## Did this PR include necessary documentation updates?
Yes

jiayuasu commented on docs/tutorial/geo_spark_core_python.md:
````diff
@@ -86,23 +86,23 @@ uses findspark Python package to upload jar files to executor and nodes. To avoi
 ```bash
 
-pipenv run python -m pip install dist/geo_pyspark-0.3.0-py3-none-any.whl
+pipenv run python -m pip install dist/geo_pyspark-1.3.0-py3-none-any.whl
````
You use geo_pyspark at some places and geo-pyspark at some other places. Can you make them consistent?

Imbruced replied:
I still do not know why PyPi gave the project name as geo-pyspark instead of geo_pyspark. That is the reason behind different names in pip install geo-pyspark and pip install from wheel file. Everywhere I am using geo_pyspark, so it is trange to me why did PyPi do that.

jiayuasu commented:
@Imbruced 

In addition to the naming convention of "geo_pyspark", another important part is that we need to put python module into our Travis-CI continuous integration. We must have this in order to ensure the correctness of the python module.

You need to add the shell script of compiling and running the python code at travis.yml. Please take a look at here: https://github.com/DataSystemsLab/GeoSpark/blob/master/.travis.yml#L22

If you don't know how to make it work, I can add it but you need to let me know how to compile and run test cases on the python codebase you have.

Imbruced commented:
I thought also about that, I will add essential bash script for Python package tests.

jiayuasu commented:
@Imbruced Thank you very much for your help!

Can your script include the test for different Spark version (Spark 2.1 and 2.2)? (or is it necessary?)

Please note that, GeoSpark for Spark 2.1 and 2.2 are maintained in two branches other than the master branch. I always cherry-pick critical updates from the master branch to Spark 2.1 and 2.2 branches.

Imbruced commented:
I will take care of that tomorrow (actually I have never tested package with Spark 2.1 and did not grant support for it, but is a good time to test that). I already have some bash scripts, which I used to test my code with Spark 2.2, 2.3, 2.4 and Python 3.6 and 3.7.

Imbruced commented:
I have template to test my package using Travis. Test time only for Python 3.6, GeoSpark 1.2.0 and Spark 2.4 lasts ~15 minutes. Should I each time run tests for other Spark versions (2.3, 2.2 and 2.1) so that means we need 15 * 4 minutes to test all Spark versions (*2 if we include python 3.7) and I do not count GeoSpark building time. Maybe should I try to decrease amount of time for complex tests ? The most time consuming are those for spatial_operator https://github.com/DataSystemsLab/GeoSpark/tree/master/python/tests/spatial_operator (>60% of time). What do you think ?  Travis configuration file which I used https://github.com/Imbruced/geo_pyspark/blob/ci_branch/.travis.yml.

jiayuasu commented:
@Imbruced 

I think for the master branch, we just test Spark 2.3 + GeoSpark 1.3 + Python 3.6 (or 3.7)

Several notes:
1. We can put the test for Spark 2.1 and 2.2 to the travis.yml in the other two branches. You can just leave this part to me.
2. For Python, is there any particular reason to test both Python 3.6 and 3.7?

In addition, I think we really should fix the issue of inconsistent python library names in PyPi. If we cannot fix it, we should change it and use a consistent name everywhere. Otherwise, this will bring the users lots of troubles in the future. I can imagine that there will be some GitHub issues about this.

More importantly, GeoSpark 1.3.0 will be a major release and we will allow significant name changes. But the new name should make sense and be able to last for a very long time.

Imbruced commented:
I had some issue with metaclasses between version 3.6 and 3.7. I read some official documents about PyPi and name should consists of lowercase letters without special charackters like "_". After quick research I can see that name geopyspark is already taken by Geomesa project. Should we change project name to <b> geospark</b> 

Furthermore I will finish Travis script today, only with Python 3.7 tests.

jiayuasu commented:
@Imbruced yes, let's change the name of GeoSpark in Python as "geospark" then

Imbruced commented:
@jiayuasu 
I will follow steps:
1. Change name in whole project, rebuild it again
2. Finish CI Travis script to test Python 3.7 with Spark 2.2 2.3 2.4 
3. Create new pull requests
4. Upload package on PyPi with new name

What do you think ?

jiayuasu commented:
@Imbruced 
This looks good to me. Just one thing, in the travis script, just keep the test on Spark 2.3.

GeoSpark master branch is compatible to both Spark 2.3 and Spark 2.4.

GeoSpark 2.1 and 2.2 branch will have their own travis script on corresponding Spark. I will do that once your PR get merged.

Imbruced commented:
@jiayuasu 
I created CI script for Python 3.7 and Spark 2.3 and 2.4. Also I updated documentation with changed name, currently package will be called **geospark**. To be sure I locked project name on PyPI with release 0.0.0 and 0.0.1, when new PR will be accepted I will push geospark to PyPI with version 1.3.0. At the moment Travis is testing latest push, once it finish I create new pull request.

Imbruced commented:
We can stay with this PR, I added -q flag to mvn clean install command, because it was failing with an error **The job exceeded the maximum log length, and has been terminated.**.

jiayuasu commented:
@Imbruced This looks good to me.

I noticed that the CI still failed because of too much log info. One  thing I can do is that: We first merge this PR and then I will remove the test output in GeoSpark SQL.

However, I wonder that how much log information you have in your python test. I am not sure whether you should also remove more log info printed by your python test.

Imbruced commented:
@jiayuasu 
Build without testing GeoSpark did not fail. All output logs from testing python is about 700 lines of log messages. I can try to remove some of them because we need only 30 free lines to make it finish without an error.

Imbruced commented:
@jiayuasu 
I removed testing Spark 2.3 because issue behind failed test was amount of time, previous exception
**The job exceeded the maximum time limit for jobs, and has been terminated.**
I think in next release I can modify tests for Python to be able to test other versions as well (limit for testing is 50 minutes). If you see any possible improvement in testing script please correct it.

jiayuasu commented:
@Imbruced This looks good to me. I saw that, on pypi, you release a geospark version 0.0.1, can you release it as 1.3.0? So the python API version will be the same as the main GeoSpark version.

Imbruced commented:
@jiayuasu 
Should I push 1.3 version on PyPI before we merge with main repo ?

jiayuasu commented:
@Imbruced Yes, please push 1.3 version to PyPi. Then I will merge this PR and make the final cut

jiayuasu commented:
@Imbruced I also notice that, you just push a single Python version for Spark version. How does it differentiate GeoSpark jars for different Spark versions?

Imbruced commented:
@jiayuasu 
While implementing GeoSpark core Python API my goal was to bring Python API for GeoSpark 1.2.0. That is the reason why I have 3 jar files built for each spark version (2.2, 2.3, 2.4). I gave easy way to extend pyspark by using function **upload_jars**. But also it is possible to add jars via Spark Config or upload them to $SPARK_HOME/jars directory. 3 jar files which I added:
<li> GeoSparkSQL </li> the same as GeoSpark SQL but with one addition for GeometryUDT (pyUDT attribute which connects python library with GeoSpark SQL)
<li> GeoSaprk Core </.li> which is the same with GeoSpark core jar file
<li> geo_wrapper </li> which has adapter classes and classes to serialize scala/java objects to ByteArray which is then deserialized to Python objects and also classes which deserialize byte arrays to scala/java objects.

For GeoSpark 1.3.0 I added change to GeoSpark SQL so there is no need to use my build jars, user can use official GeoSpark SQL jar with GeoSparkCore and geo_wrapper (which is only required for GeoSpark core Python).

If user wants to use GeoSpark Core, he has to also add geo_wrapper jar file. We can merge my scala code to GeoSpark core code so there will be no need to add additional jar file (geo_wrapper), that may make some confusion (or we merge that in GeoSpark 1.3.1), please let me know what you think.

For previous GeoSpark versions I have built jar files, in directory python/geospark/jars/previous. By default it is newest version (directories 2_3, 2_4, 2_2). I have implemented decorators which inform user since which version function is available (ex WKTReader class is available since GeoSpark 1.2.0). In wheel file I am packing only 2_2, 2_3, 2_4 other jars can be found on repo. I hope that answers your question.

jiayuasu commented:
@Imbruced Thank you for the explanation!

Please release GeoSpark 1.3.0 on Pypi. Then I will merge this PR😁

Imbruced commented:
Package is officially available with version 1.3.0 on PyPi https://pypi.org/project/geospark/. 
Also I updated previous jars version for GeoSpark SQL :)

Imbruced commented:
I removed previous project from PyPi (geo-pyspark). Currently **geospark** is only one which is available.

jiayuasu commented:
@Imbruced LGTM! Thank you very much!

jiayuasu commented:
@Imbruced Sorry, I just found one thing:

In this PR, you also changed the content of GeometryUDT.scala "py_udt_string", right? However, the udt string with "geo_pyspark" has been released to Maven Central as GeoSpark 1.3.0. Will this cause a problem?

Imbruced commented:
All GeoSparkSQL functions will work for Python API but it will fail with converting to shapely objects (methods such as toPandas, collect etc.). My packed jars does not have that issue (geospark/jars...).
Is there any way to update that on Maven or is the same as on PyPI (only by adding post release) ?

jiayuasu commented:
@Imbruced There is no way to update a released jar on Maven Central. The only way to fix it is to release GeoSpark 1.3.1. Otherwise, GeoSpark Maven JAR won't work with python API.

My bad, I shouldn't release GeoSpark Jars on Maven before this PR.

The packed jar in the testing script uses the updated GeometryUDT.scala. So it didn't show the error previously.

OK, my plan is to (1) I will release GeoSpark 1.3.1 on Maven right away (2) You will release GeoSpark Python 1.3.1 on Pypi. Do you need to change version info in your python code to make it work with GeoSpark 1.3.1?

Imbruced commented:
I do not have to change version number to make it work with GeoSpark 1.3.1 but I have to add that to build wheel file with version 1.3.1 which are essential on PyPi. In case of python https://github.com/DataSystemsLab/GeoSpark/blob/master/python/geospark/version.py version variable has to be changed. If possible please update also mine version. I will push 1.3.1 version on PyPI.

jiayuasu commented:
@Imbruced 

GeoSpark 1.3.1 has been released. I also updated https://github.com/DataSystemsLab/GeoSpark/blob/master/python/geospark/version.py

Please update PyPi on your side and then we are good to go.

Imbruced commented:
Done ;) On PyPi we have geospark 1.3.1
